log/experimental_level/level.go

```diff
@@ -0,0 +1,146 @@
+// Its return value may be provided as the Allowed parameter in the Config.
+func AllowErrorOnly() []string {
+	return []string{errorLevelValue}
+}
+
+// AllowNone allows none of the default log levels.
+// Its return value may be provided as the Allowed parameter in the Config.
+func AllowNone() []string {
+	return []string{}
+}
+
+// Error returns a logger with the level key set to ErrorLevelValue.
+func Error(logger log.Logger) log.Logger {
+	return log.NewContext(logger).With(levelKey, errorLevelValue)
+}
+
+// Warn returns a logger with the level key set to WarnLevelValue.
+func Warn(logger log.Logger) log.Logger {
+	return log.NewContext(logger).With(levelKey, warnLevelValue)
+}
+
+// Info returns a logger with the level key set to InfoLevelValue.
+func Info(logger log.Logger) log.Logger {
+	return log.NewContext(logger).With(levelKey, infoLevelValue)
+}
+
+// Debug returns a logger with the level key set to DebugLevelValue.
+func Debug(logger log.Logger) log.Logger {
+	return log.NewContext(logger).With(levelKey, debugLevelValue)
+}
+
+// Config parameterizes the leveled logger.
+type Config struct {
+	// Allowed enumerates the accepted log levels. If a log event is encountered
+	// with a level key set to a value that isn't explicitly allowed, the event
+	// will be squelched, and ErrNotAllowed returned.
+	Allowed []string
+
+	// ErrNotAllowed is returned to the caller when Log is invoked with a level
+	// key that hasn't been explicitly allowed. By default, ErrNotAllowed is
+	// nil; in this case, the log event is squelched with no error.
+	ErrNotAllowed error
+
+	// SquelchNoLevel will squelch log events with no level key, so that they
+	// don't proceed through to the wrapped logger. If SquelchNoLevel is set to
+	// true and a log event is squelched in this way, ErrNoLevel is returned to
+	// the caller.
+	SquelchNoLevel bool
+
+	// ErrNoLevel is returned to the caller when SquelchNoLevel is true, and Log
+	// is invoked without a level key. By default, ErrNoLevel is nil; in this
+	// case, the log event is squelched with no error.
+	ErrNoLevel error
+}
+
+// New wraps the logger and implements level checking. See the commentary on the
+// Config object for a detailed description of how to configure levels.
+func New(next log.Logger, config Config) log.Logger {
+	return &logger{
+		next:           next,
+		allowed:        makeSet(config.Allowed),
+		errNotAllowed:  config.ErrNotAllowed,
+		squelchNoLevel: config.SquelchNoLevel,
+		errNoLevel:     config.ErrNoLevel,
+	}
+}
+
+type logger struct {
+	next           log.Logger
+	allowed        map[string]struct{}
+	errNotAllowed  error
+	squelchNoLevel bool
+	errNoLevel     error
+}
+
+func (l *logger) Log(keyvals ...interface{}) error {
+	var hasLevel, levelAllowed bool
+	for i := 0; i < len(keyvals); i += 2 {
+		if k, ok := keyvals[i].(string); !ok || k != levelKey {
+			continue
+		}
+		hasLevel = true
+		if i >= len(keyvals) {
```

Comment:
What is this check for? Unless I'm missing something I believe this is always false.